### PR TITLE
Add evm key store

### PR DESCRIPTION
**Pull request by cryptoAtwill**
Add evm key store. This implementation follows a similar approach as that in the `identity` crate. Changes:
- Add traits for key store operations
- Add memory key store
- Add persistence key store, without encryption.

### CODE REVIEW

File: eth-keystore/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
+
+[dependencies]
+zeroize = "1.6.0"
+anyhow = { workspace = true }
+ethers = { workspace = true, optional = true }
+serde = { workspace = true }
+serde_json = { workspace = true }
+hex = "0.4.3"
```
Comment (adlrocha) [inline eth-keystore/Cargo.toml:14]: Should we make this dependency part of the workspace? IIRC we are using it in other crates also.

File: eth-keystore/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+/// The key store trait for different evm key store
+pub trait KeyStore {
+    /// The type of the key that is stored
+    type Key: Clone + Eq + Hash + TryFrom<KeyInfo>;
+
+    /// Get the key info by address string
+    fn get(&self, addr: &Self::Key) -> Result<Option<KeyInfo>>;
+    /// List all addresses in the key store
+    fn list_all(&self) -> Result<Vec<Self::Key>>;
```
Comment (adlrocha) [inline eth-keystore/src/lib.rs:23]: We are calling this function `list` in the Filecoin keystore, we should maybe use the same names for consistency?

File: eth-keystore/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+/// The struct that contains evm private key info
+#[derive(Clone)]
+pub struct KeyInfo {
+    private_key: Vec<u8>,
+}
+
+impl Drop for KeyInfo {
+    fn drop(&mut self) {
+        self.private_key.zeroize();
```
Comment (adlrocha) [inline eth-keystore/src/lib.rs:36]: This is to avoid having the private key set in-memory, right?

File: eth-keystore/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    }
+}
+
+#[cfg(feature = "with-ethers")]
```
Comment (adlrocha) [inline eth-keystore/src/lib.rs:40]: Why do we need this feature?

File: eth-keystore/src/persistent.rs [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+        }
+
+        Ok(Self {
+            memory: MemoryKeyStore { data: key_infos },
+            file_path: path,
+        })
+    }
+
+    /// Write all keys to file without any encryption.
+    fn write_all_no_encryption(&self) -> Result<()> {
```
Comment (adlrocha) [inline eth-keystore/src/persistent.rs:87]: Should we call this function `flush_no_encryption`?

File: eth-keystore/src/persistent.rs [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+
+        fs::create_dir_all(dir)?;
+
+        let file = File::create(&self.file_path)?;
+
+        // TODO: do we need to set path permission?
+
+        let writer = BufWriter::new(file);
+
+        let to_persist = self
```
Comment (adlrocha) [inline eth-keystore/src/persistent.rs:101]: It may be a bit inefficient to re-write all keys every time that we put a new identity, right? We should at least track a `TODO` here describing that this can probably be improved.

Also, we should double-check in tests that we don't write duplicates every time that we put new keys.

File: eth-keystore/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+    /// The type of the key that is stored
+    type Key: Clone + Eq + Hash + TryFrom<KeyInfo>;
+
+    /// Get the key info by address string
+    fn get(&self, addr: &Self::Key) -> Result<Option<KeyInfo>>;
+    /// List all addresses in the key store
+    fn list_all(&self) -> Result<Vec<Self::Key>>;
+    /// Put a new info to the addr
+    fn put(&mut self, info: KeyInfo) -> Result<()>;
+}
```
Comment (adlrocha) [inline eth-keystore/src/lib.rs:26]: We are missing a remove function. We will need it so I suggest including it already.
```suggestion
    fn remove(&mut self, addr: &Self::Key)
}
```

**Review by adlrocha — CHANGES_REQUESTED** [review]
Overall LGTM,  but we should introduce a `remove` function. I also left a few other comments/suggestions. We can discuss sync if needed.

File: eth-keystore/src/lib.rs (hunk 4 above)
Comment (cryptoAtwill) [reply eth-keystore/src/lib.rs:40]: The key to the key store does not assume the type of the key, so we dont have a hard dependency on `ethers`. So that we can expand to other crates, just more options and dont force downstream to use ethers.

File: Cargo.toml [hunk 8]
```diff
@@ -1,5 +1,5 @@
 [workspace]
-members = [".", "testing/e2e", "identity", "testing/itest"]
+members = [".", "testing/e2e", "identity", "testing/itest", "eth-keystore"]
```
[the remark this answers is not in the document]
Comment (cryptoAtwill) [reply Cargo.toml:2]: yeah, initially I was thinking along this line, but when I started to implement it, I realised it would make the `identity` crate quite messy or need some refactoring, like renaming some struct and move around the files. So I started with a new crate.

File: eth-keystore/src/lib.rs (hunk 3 above)
Comment (adlrocha) [reply eth-keystore/src/lib.rs:36]: I like it!

File: Cargo.toml (hunk 8 above)
[the remark this answers is not in the document]
Comment (adlrocha) [reply Cargo.toml:2]: I see... I think it is worth doing the effort now of following that approach, it may be harder to refactor in the future, WDYT?

[the remark this answers is not in the document]
Comment (cryptoAtwill) [reply Cargo.toml:2]: this works, time to see how good the IDE is.

File: eth-keystore/src/persistent.rs (hunk 6 above)
Comment (cryptoAtwill) [reply eth-keystore/src/persistent.rs:126]: I followed the implementation of `identity`, copied and adapted the tests over.

**Review by adlrocha — APPROVED** [review]
